Add ViralConsensus v0.0.4

niemasd opened the pull request:
This PR adds ViralConsensus v0.0.3 to bioconda.

----

Please read the [guidelines for Bioconda recipes](https://bioconda.github.io/contributor/guidelines.html) before opening a pull request (PR).

### General instructions

* If this PR adds or updates a recipe, use "Add" or "Update" appropriately as the first word in its title.
* New recipes not directly relevant to the biological sciences need to be submitted to the [conda-forge channel](https://conda-forge.org/docs/) instead of Bioconda.
* PRs require reviews prior to being merged. Once your PR is passing tests and ready to be merged, please issue the `@BiocondaBot please add label` command.
* Please post questions [on Gitter](https://gitter.im/bioconda/Lobby) or ping `@bioconda/core` in a comment.

### Instructions for avoiding API, ABI, and CLI breakage issues
Conda is able to record and lock (a.k.a. pin) dependency versions used at build time of other recipes.
This way, one can avoid that expectations of a downstream recipe with regards to API, ABI, or CLI are violated by later changes in the recipe.
If not already present in the meta.yaml, make sure to specify `run_exports` (see [here](https://bioconda.github.io/contributor/linting.html#missing-run-exports) for the rationale and comprehensive explanation).
Add a `run_exports` section like this:

```yaml
build:
  run_exports:
    - ...

```

with `...` being one of:

| Case                             | run_exports statement                                               |
| -------------------------------- | ------------------------------------------------------------------- |
| semantic versioning              | `{{ pin_subpackage("myrecipe", max_pin="x") }}`     |
| semantic versioning (0.x.x)      | `{{ pin_subpackage("myrecipe", max_pin="x.x") }}`   |
| known breakage in minor versions | `{{ pin_subpackage("myrecipe", max_pin="x.x") }}` (in such a case, please add a note that shortly mentions your evidence for that) |
| known breakage in patch versions | `{{ pin_subpackage("myrecipe", max_pin="x.x.x") }}` (in such a case, please add a note that shortly mentions your evidence for that) |
| calendar versioning              | `{{ pin_subpackage("myrecipe", max_pin=None) }}`    |

while replacing `"myrecipe"` with either `name` if a `name|lower` variable is defined in your recipe or with the lowercase name of the package in quotes.

### Bot commands for PR management

<details>
  <summary>Please use the following BiocondaBot commands:</summary>

Everyone has access to the following BiocondaBot commands, which can be given in a comment:

<table>
  <tr>
    <td><code>@BiocondaBot please update</code></td>
    <td>Merge the master branch into a PR.</td>
  </tr>
  <tr>
    <td><code>@BiocondaBot please add label</code></td>
    <td>Add the <code>please review & merge</code> label.</td>
  </tr>
  <tr>
    <td><code>@BiocondaBot please fetch artifacts</code></td>
    <td>Post links to CI-built packages/containers. <br />You can use this to test packages locally.</td>
  </tr>
</table>

Note that the <code>@BiocondaBot please merge</code> command is now depreciated. Please just squash and merge instead.

Also, the bot watches for comments from non-members that include `@bioconda/<team>` and will automatically re-post them to notify the addressed `<team>`.

</details>

niemasd commented:
@BiocondaBot please add label

jmarshall commented:
Your _meta.yaml_ file lists your package as depending on bioconda's htslib package. This is the right way to build a bioconda package: depend on the common externally packaged library, rather than bundle it within your package.

Corresponding to this, in _build.sh_ you should remove the cloning and building of htslib and instead patch and/or configure ViralConsensus (if necessary) so that it can build against an already installed htslib.

niemasd commented:
@jmarshall Thanks for the feedback! I've refactored the ViralConsensus code base to use the environment's htslib installation, and I've refactored the bioconda recipe to properly use the conda environment's htslib. It now passes the tests again. Can you review again?

jmarshall commented on recipes/viral_consensus/build.sh:
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
+
+# update environment variables
+export CFLAGS="$CFLAGS -I$PREFIX/include"
+export CPPFLAGS="$CPPFLAGS -I$PREFIX/include"
+export LDFLAGS="$LDFLAGS -L$PREFIX/lib"
+export CPATH=${PREFIX}/include
+export INCLUDES="-I${PREFIX}/include"
+export LIBPATH="-L${PREFIX}/lib"
+export LIBRARY_PATH=${PREFIX}/lib
```
There's no need to define any of these, as you're using conda's compilers. (It's possible that might break and you might need to re-add one or two of these, but I expect not.)

niemasd replied:
Thanks! I'll remove them, and if something breaks, I'll re-add them as needed (I think the Mac OS X build was failing until I added `export LIBRARY_PATH=${PREFIX}/lib`)

jmarshall commented on recipes/viral_consensus/build.sh:
```diff
@@ -0,0 +1,18 @@
+make CC=$CC CXX="${CXX}"
+if [ ! -d "$PREFIX/bin" ]; then
+    mkdir $PREFIX/bin;
+    export PATH=$PREFIX/bin:$PATH;
```
No need to add to PATH for _build.sh_.
Most recipes just use `mkdir -p $PREFIX/bin` instead of a conditional.

niemasd replied:
Thanks! I'll replace with just `mkdir -p` and remove the conditional

jmarshall commented on recipes/viral_consensus/meta.yaml:
```diff
@@ -0,0 +1,47 @@
+
+source:
+  url: https://github.com/niemasd/ViralConsensus/archive/refs/tags/{{ version }}.tar.gz
+  sha256: d531311c8e5c6b689f4426a9e89cc00426733ae9d4ff71757a861ad6ddbf81f4 # curl -sL URL | openssl sha256
```
In future the autobump script will show you new sha256 values, so you won't need the mnemonic comment.

niemasd replied:
Thanks! I more so left it as a note for myself if I ever add more tools to bioconda in the future, but I'll remove it for cleanliness (it was easy to Google anyways)

jmarshall commented on recipes/viral_consensus/meta.yaml:
```diff
@@ -0,0 +1,47 @@
+
+requirements:
+  build:
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+    - bzip2
+    - htslib
+    - libcurl
+    - make
+    - ncurses
+    - zlib
```
Just `make` and the two compilers for `build:`

niemasd replied:
Thanks! Removed everything else

jmarshall commented on recipes/viral_consensus/meta.yaml:
```diff
@@ -0,0 +1,47 @@
+    - ncurses
+    - zlib
+  host:
+    - bzip2
+    - htslib
+    - libcurl
+    - ncurses
+    - zlib
```
Just `htslib` for `host:` — you don't use `ncurses` at all and the others you don't use directly.

niemasd replied:
Thanks! Removed everything else

jmarshall commented on recipes/viral_consensus/meta.yaml:
```diff
@@ -0,0 +1,47 @@
+    - ncurses
+    - zlib
+  run:
+    - bzip2
+    - htslib
+    - libcurl
+    - ncurses
+    - zlib
```
Just `htslib` for `run:` or in fact you should probably leave this empty and the dependency will be automatically derived.

niemasd replied:
Thanks! A similar tool (`ivar`) seems to have `htslib` under `run:`, but I'll try it without it and re-add if it breaks things

jmarshall replied:
In the past it was necessary, but more recently most libraries such as htslib propagate the right thing from `host:`.

niemasd commented:
Following up after the tests finished: Linux is crash due to not being able to find the `-llzma`, `-lbz2`, `-lz`, `-lcurl`, and `-lhts` libraries. Not sure if it's because I removed these from `build:` in `meta.yaml` or because of the removal of environmental variable assignments in `build.sh`. Will play around and find out

jmarshall commented:
Make sure `-lhts` appears before the others.

niemasd commented:
As in edit the `Makefile` in my tool's GitHub repo to move `-lhts` earlier than `-llzma -lbz2 -lz -lcurl -pthread` in the `LIBS` declaration? I could do that, but I'd need to make a new release for it (otherwise the updated `Makefile` isn't in the current tarball)

niemasd commented:
@jmarshall I just rereleased v0.0.4 with `-lhts` first in the `LIBS` definition in the `Makefile`, and I updated the bioconda `meta.yaml` with the updated `sha256`. The tests are running, so fingers crossed it's okay to merge this time! 😄

niemasd commented:
@jmarshall Strange, just making this single change (moving `-lhts` first in the `Makefile` definition of `LIBS`) broke the Mac OS X build:

https://github.com/niemasd/ViralConsensus/commit/1ebe1584a50048fb5f68e7a8a96643518c305617

I'm not quite sure how to debug, as I don't have a Mac environment, and I can't seem to tell what failed in the logs:

https://dev.azure.com/bioconda/bioconda-recipes/_build/results?buildId=42408&view=logs&j=1b052f1d-4456-52f0-9d43-71c4c5bd734d&t=edc48dcd-1fc2-5e3b-7036-7be9cea00123&l=353

jmarshall commented:
This is likely something that I meant to mention previously: you should propagate conda's CXXFLAGS etc settings to your Makefile too. So in _build.sh_:

```sh
make CC="$CC" CXX="$CXX" CFLAGS="$CFLAGS" CPPFLAGS="$CPPFLAGS" CXXFLAGS="$CXXFLAGS" LDFLAGS="$LDFLAGS"
```

I see you had added some others too, but with the full complement of these standard ones you shouldn't need manual ones.

niemasd commented:
@jmarshall That now causes the Linux build to fail as well, for the same `ld` error message of not being able to find the libraries (I think because the library paths are not being set). I'll revert some of the changes until it gets back to a working state, and I'd rather submit it once it's back to a working state if possible, as I'm not sure how much cleaner it can get (it seems like it's already quite a bit cleaner than many other recipes in bioconda, and this is taking quite some time to tinker with)

niemasd commented:
@jmarshall After re-adding those library paths to the `make` command variables, it seems to have worked again and is now passing all the tests 😄

jmarshall commented:
The error messages are what tell us what is wrong with your recipe. I will be examining the previous build logs further.

It would not be appropriate to merge this in its current state, without understanding the causes of the various failures.

niemasd commented:
Sure, makes sense; thanks for checking the build logs!

jmarshall commented:
Unsurprisingly, the problem is that your Makefile doesn't actually use those conventional variables. I've pushed a commit to have the recipe put conda's standard settings variables into the `$(RELEASEFLAGS)` variable that the Makefile does use.

This is better than papering over the cracks with explicit `-L` options and `$LIBRARY_PATH` environment variables, as it preserves conda's `-rpath` settings that presumably are important to producing properly usable packaged executables (though you might expect that being missing to cause the test step's trial execution to fail too, which appears not to have happened).

jmarshall commented:
@BiocondaBot please fetch artifacts

niemasd commented:
Thank you!!

jmarshall commented:
No problem :smile:

Bit annoying that the system has uploaded one of the already superseded 0.0.3 versions, but it won't really do any harm.
